DSC resource Unit/Template test initialization/tear down functions added

PlagueHO opened the pull request:
As discussed in Issue #45 in the xNetworking resource (https://github.com/PowerShell/xNetworking/issues/45).

This adds two new functions to TestHelpers.psm1 (the functions contain help and example usage):
Initialize-TestEnvironment
Restore-TestEnvironment

I've also added the Unit and Integration test templates from the xNetworking module to this repo as well (as they are generic and ideally used with any DSC resource). These templates use the above functions to set up and tear down the environment.

I have tested this with my fork of the xNetworking repo where I updated the tests to use this helper code instead (see this here: https://github.com/PlagueHO/xNetworking)

@tysonjhayes, would you mind taking a look at this when you get a chance? It is pretty much the code we've been using in the xNetworking tests. I've also updated my xNetworking fork to use this code as well so it can be PR'd once this PR has gone in.

Cheers!

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+        $null = Copy-Item -Path $TestEnvironment.BackupLocation -Destination "${env:ProgramFiles}\WindowsPowerShell\Modules" -Recurse -Force
+        Remove-Item -Path $TestEnvironment.BackupLocation -Recurse -Force
+    }  
+}
```
You'll want to add an extra line here.

PlagueHO replied:
I've added one at the end here, but it's not displayed above for some reason.

tysonjhayes commented on Templates/integration_config_template.ps1:
```diff
@@ -0,0 +1,24 @@
+#>
+
+
+# TODO: Modify ResourceName
```
We should add in a dot source to the 'MSFT_<xResourceName>'.ps1 to load in the config data for the integration tests.

tysonjhayes commented on Templates/integration_template.ps1:
```diff
@@ -0,0 +1,74 @@
+# /TODO
+
+#region HEADER
+Import-Module .\DSCResource.Tests\TestHelper.psm1 -Force
```
I don't love this to be honest, wouldn't this mean we have to have a DSCResource.Tests in the unit\integration folder in order to properly import this if we took the template whole sale?

Thinking around this I see a few options.

1) Remove the import-module and put in a comment where to download/clone DSCResource.Tests - this puts the onus on the test runner to setup the system before the tests run successfully. It also means we don't automatically setup the system for them which may not be cool to some people and has been something we've avoided in the past.

2) We try to automatically setup the system to run the tests, this means we'd have to clone/install DscResource.Tests ourselves, this would add more overhead to our code and would mean checking/installing during every integration/unit test.  Also means our git repos get dirtied up with a extra cruft if we put it directly in the folder so we'd probably want to remove the file when we're done.

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+function Initialize-TestEnvironment {
+    [OutputType([PSObject])]
+    [CmdletBinding()]
+    Param (
```
Please put ( on it's own line please.

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+        [String] $DSCResourceName,
+
+        [ValidateSet('Unit','Integration')]
+        $TestType = 'Unit'
```
Should probably be [String] $TestType to match the others.

Also do we want to set Unit as the default? Ideally we should be checking in with both unit/integration tests so we do want to be giving priority to one?

PlagueHO replied:
Agreed. Makes it clearer too. I'll remove the default.

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+    if(-not (Test-Path -Path $moduleRoot))
+    {
+        $null = New-Item -Path $moduleRoot -ItemType Directory
+        [String] $BackupLocation = ''
```
We should probably set a style guide on how we want to deal with empty strings. Personally I find [string]::empty more eye catching (ie:
````powershell
[string] $BackupLocation = [string]::Empty
````

However it doesn't actually make a difference besides readability.

PlagueHO replied:
That is more readable (or at least much more obvious that it is empty). A style guide in general for all DSC would be really helpful. Could a section be added to covering style:

https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md

I generally don't have a particular style preference myself - so I'm happy to go with what ever. But I am a bit fussy over consistency so a style guide is key for that. Some aspects of style could actually be tested by the Meta.Tests.ps1 as well should they be important enough (but that is a separate discussion).

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+        This command will initialize the test enviroment for Unit testing
+        the MSFT_xFirewall DSC resource in the xNetworking DSC module.      
+#>
+function Restore-TestEnvironment  {
```
Curly brace on it's own line please.

tysonjhayes commented on TestHelper.psm1:
```diff
@@ -201,3 +201,216 @@ function Install-ResourceDesigner {
+#>
+function Restore-TestEnvironment  {
+    [CmdletBinding()]
+    Param (
```
Paraenthesis on it's own line please.

tysonjhayes commented:
So my comment doesn't get lost in the diff changes.

I'm basically debating the logic of where/how we should be calling these functions. Personally I don't even have DscResource.Tests on my machine as it was always a pain to remember to call it before submitting. With this template change as resources start adopting it test runners will have to have the module and keep it updated (or else when we make changes to this it'll fail on AppVoyer which is always running current).

That being said it's the easiest way to drop in the pre/post logic in an easily updatable way that would keep consistence. So big advantage there.  I'll have to think on this some more.

@KarolKaczmarek and @TravisEz13  what is your opinion on the logic here?

PlagueHO commented:
I completely understand what you're saying here. Ideally I would have preferred the code from this repo to be downloaded automatically via PowerShellGallery or something like that. But the content of this repo must always be put in a subfolder of the DSC Module being tested (which is automatically done via AppVeyor normally) so that Pester will always fires the Meta.Test.ps1 file on the DSC.

I have found it useful in my projects to make sure I put this folder into them before I test them locally, because I have committed code a number of times that has failed because AppVeyor does execute these tests and my local tests didn't. I guess in those situations I'd always rather find out before I commit - which is why I've ensured the tests fail with an instructional error message telling the user what to do if the code isn't there.

But I know you're right - there must be a better way of doing this.

In the meantime I'll fix up the other stuff. Thanks heaps for looking at this!

tysonjhayes commented on Templates/integration_template.ps1:
```diff
@@ -0,0 +1,83 @@
+# /TODO
+
+#region HEADER
+if ( (-not (Test-Path -Path '.\DSCResource.Tests\')) -or `
```
I think instead of doing this we could do a Get-Module DSCResource.Tests and import if it exists, and if it doesn't then throw this. This way we are not demanding that we have a clone of the files in the repo but can still excercise the function.

KarolKaczmarek commented on Templates/integration_template.ps1:
```diff
@@ -0,0 +1,83 @@
+            {
+                [System.Environment]::SetEnvironmentVariable('PSModulePath',
+                    $env:PSModulePath,[System.EnvironmentVariableTarget]::Machine)
+                Invoke-Expression -Command "$($DSCResourceName)_Config -OutputPath `$TestEnvironment.WorkingFolder"
```
$($DSCResourceName)_Config 
is this correct? The configuration in integration_config_template.ps1 seems to be called simply $DSCResourceName

PlagueHO replied:
Good catch! I've changed the config name in the config template to match this rather than change the code here so that it is more obvious what is being called.

KarolKaczmarek commented on Templates/integration_template.ps1:
```diff
@@ -0,0 +1,83 @@
+   DSCResource.Tools in Future and therefore should not be altered if possible.
+#>
+
+# TODO: Customize these paramters...
```
just a typo in "param**e**ters"

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,221 @@ function Install-ResourceDesigner {
+
+    .OUTPUT
+        Returns a test environment object which must be passed to the
+        Restore-TestEnvironment function to allow it to restore the system
```
typo

PlagueHO replied:
Fixed.

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,221 @@ function Install-ResourceDesigner {
+        back to the original state as well as clean up and working/temp files.
+        
+    .EXAMPLE
+        $TestEnvrionment = Inialize-TestEnvironment `
```
typo in TestEnvrionment

KarolKaczmarek commented on Templates/integration_template.ps1:
```diff
@@ -0,0 +1,83 @@
+                [System.Environment]::SetEnvironmentVariable('PSModulePath',
+                    $env:PSModulePath,[System.EnvironmentVariableTarget]::Machine)
+                Invoke-Expression -Command "$($DSCResourceName)_Config -OutputPath `$TestEnvironment.WorkingFolder"
+                Start-DscConfiguration -Path (Join-Path -Path $env:Temp -ChildPath $DSCResourceName) `
```
Above you use $TestEnvironment.WorkingFolder, here (Join-Path -Path $env:Temp -ChildPath $DSCResourceName)
I can see that it eventually evaluates to the same thing, but can be confusing when someone reads it first time and is not familiar with Initialize-TestEnvironment.
I would recommend using $TestEnvironment.WorkingFolder here as well.

PlagueHO replied:
Fixed.

KarolKaczmarek commented on Templates/unit_template.ps1:
```diff
@@ -0,0 +1,84 @@
+
+    #region Pester Tests
+
+    InModuleScope $DSCResourceName {
```
You may want to add comment on what InModuleScope is as it's a bit more advanced concept and not everyone using the template will be familiar with it.

PlagueHO replied:
Done. I just copied the text from the Pester Wiki (as it is easy to understand I think).

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,221 @@ function Install-ResourceDesigner {
+    
+    # If this module already exists in the Modules folder, make a copy of it in
+    # the temporary folder so that it isn't accidentally used in this test.
+    if(-not (Test-Path -Path $moduleRoot))
```
I'm not sure we should be removing, even temporarily any modules which are already on user's machine.
I think the simple, non-invading solution would be to place module to be tested in some temporary location and add that location to the beginning of $env:PsModulePath until we finish testing.
What do you guys think?

PlagueHO replied:
I agree - plus it eliminates the issue of the tests terminating and leaving the modules in an unknown state. That does sound like a pretty solution to me. But I'd definitely want @tysonjhayes 's opinion on this one.

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,221 @@ function Install-ResourceDesigner {
+
+    # Temp Working Folder - always gets removed on completion
+    # The tests can put anything in here and it will get cleaned up.
+    [String] $WorkingFolder = Join-Path -Path $env:Temp -ChildPath $DSCResourceName
```
Can we also create this folder here so that tests can safely assume it exists?
Let's also add some random number to name of this folder so that we don't have any collisions/remove someone's existing data.

PlagueHO replied:
Good idea. Done.

vors commented:
Hi @PlagueHO thank you for thoughtful work on this issue!
The purpose of DscResource.Tests is narrow: it guarantees some level of consistency and correctness for DSC resources. It's questionable what **some** means. 
Currently it is:
- encoding
- indention
- scripts-parsability
- coherence of mof schema and function parameters in resource scripts

Key properties of all tests in this repo is: user can **avoid getting this repo locally** to fix problems reported by CI. If it's not true, then we need a better error messages. 

We have fixers, but, realistically you can convert encoding and such in your favorite text editor, as far as error message is clear.

Non-trivial logic in this repo will make a need in it local presents un-avoidable.
It's better to keep in on per-repo basis, so maintainers has a granular control over what kind of tests they want to use and such.

A good place for template and setup-teardown script is https://github.com/PowerShell/DscResources as discussed in https://github.com/PowerShell/DscResources/issues/9

Thank you again for the thoughtful PR.

KarolKaczmarek commented:
I think the main advantage of using these templates is that users will run the same tests on their machines as our CI system (including DscResource.Tests). This will help catch many bugs before PR is sent and tests executed on appveyor.
Maintainers can still opt to not use the template , so they aren't deprived of any control. 

Regarding @tysonjhayes question on obtaining this repo, I think it's fine to just pull it from github to get the most up to date version.

My main concern is whether we should be should be removing, even temporarily any modules which are already on user's machine. I stated it in comments but they will soon disappearing so writing it here as well.
I think the simple, non-invading solution would be to place module to be tested in some temporary location and add that location to the beginning of $env:PsModulePath until we finish testing.
 What do you guys think?

Also, after adding this we should modify https://github.com/PowerShell/DscResources/blob/master/CONTRIBUTING.md#adding-test-coverage-for-dsc-resources so that it recommends use of these templates.

PlagueHO commented:
Good info there @vors - I didn't realize that a lot of this had been discussed before. I'm definitely in agreement with @KarolKaczmarek about avoiding moving modules around - we want to avoid this if there is a better way. Besides, in the test initialization we're modifying the modules path anyway.

As for the other stuff, there are two questions here that I can see:
1. Should the templates stay in the DSCResource.Tests module or move to https://github.com/PowerShell/DscResources?
2. Should the test framework initialize/tear down functions (Initialize-TestEnvironment, Restore-TestEnvironment) remain in the **TestHelper** module or move into another repo (DSCResources or a new one completely)?

For 1:
It might make sense to move these templates over to DSCResources, but I don't really have an opinion either way - as long as their are instructions telling someone what they should do when starting a new resource.

For 2:
Keeping these functions out of DSCResources is a good idea, because then they could be updated/improved in future (non breaking improvements of course). If they get moved into DSCResources then they're pretty much fixed as you'd expect any newly created resource to include a copy of these functions as they were at the time of creating the new resource. I actually think keeping these functions in the TestHelper module makes the most sense because this repo is always automatically pulled by CI anyway so that the Meta.tests.ps1 automatically runs.

If these functions are moved into a new repo/module so that they can be pulled via Install-Module off PowerShell gallery, we still run into the problem where Install-Module isn't available on WMF4.0 machines that haven't specifically had the PowerShellGet module installed.

An idea @tysonjhayes :
Is it possible to cause the DSCResource.Tests resource to be included as a Git Submodule of any DSC Resources that the templates are being used with (https://git-scm.com/book/en/v2/Git-Tools-Submodules). I've not used this myself, but it sounds like it might solve the issue with forcing a user to pull this repo into their DSC resource repo for the tests to work. It could also make sure they are using the latest test helpers. If this sounds like it might be a possibility I'm happy to look into it further. It may also eliminate the need to have the specific call in AppVeyor to pull this repo.

Thanks all for your feed back on this! :smile:

KarolKaczmarek commented:
Regarding using submodules, I can only say it would not work due to the way submodules are implemented. When adding submodule, we reference the specific commit and it's hardcoded (you can see https://github.com/PowerShell/DscResources/tree/master/xDscResources as we use submodules there). The problem with that is that if there are changes in original submodule's repo, our submodule does not automatically get updated, but always refer the commit which was most current when repo was added. We would need to either manually update what commit our submodule refers to whenever there are new changes or after cloning repo, we would need to pull new changes.
In this regards, submodules are implemented really poorly in git :(

PlagueHO commented:
Right, I think I understand: any repo that used this as a submodule would need to know to issue a ```git submodule update --remote``` command to update to the latest version of the submodule.

So the choices are either pull this repo manually or pull a new module using Install-Module.

I tend to lean towards pulling the repo because it contains other files (meta.test.ps1) that **must** be in a subfolder of the DSC resource - otherwise those tests aren't run and there is a possibility the CI tests will fail where as the local tests passed.

It does mean an extra step is required before you can begin working on a DSC resource though. It would be nice if it was automatic. The templates could always call GIT itself if the folder is missing rather than throw an error - GIT should always be available if you're working on a Github repo anyway. Plus calling GIT directly eliminates the problem of the PowerShellGet module not being available on WMF4.

vors commented:
@PlagueHO yes, I think templates should leave in DscResources central repo, together with  documentation.

PlagueHO commented:
Cool. So in Summary:
1. I'll submit a PR with the templates + documentation (once all finalized) to DscResources central repo and remove templates from this PR. Perhaps it might be a good idea to include a .gitignore file in the DSCResources central repo containing the DSCResource.Tests folder so this doesn't get commited by accident to any other projects.

2. Change Initialize-TestEnvironment/Restore-TestEnvironment so that instead of moving modules to temp location, the $ENV:PSModulePath is modified to ensure loading of tested module first.

3. Change the templates so that rather than alerting the user that the DSCResource.Test folder is missing, it will use GIT to automatically pull the repo. It could be wrapped in a ShouldProcess to allow the user to confirm.

I know you US chaps are off celebrating thanksgiving, so I'll wait till next week before I make the above changes to give you a chance to weight in :smile: 

Thanks!

PlagueHO commented:
I've completed the above changes to this PR:
1. Removed the Test Templates from this repo and moved them into my fork of the DSCResources repo (I'll PR this once this PR has gone through).
2. Updated the Initialize-TestEnvironment/Restore-TestEnvironment so that they no longer move the module to the Modules folder but test the resource/modules in place.

I have also made the following changes to other Repos:
1. In my fork of [DSCResources](https://github.com/PlagueHO/DscResources):
 - added the Test templates to the DSCResources repo
 - modified the Test templates to use git to automatically pull the DSCResource.Tests repo if missing.
 - added a TestingGuidelines.md file
 - added a .gitignore file to the DSCResource.Template folder containing the entry DSCResource.tools (so that any new projects won't accidentally commit this).

2. In my fork of [xNetworking](https://github.com/PlagueHO/xNetworking):
 - changed all Integration and Unit tests to match the new templates
 - changed all Integration and Unit tests to use the Test helper code in DSCResource.Tests

I won't submit PR's for the above changes until this one is through.

Thanks!

PlagueHO commented:
Any thoughts on this one chaps?

I've been using my fork of this for a couple of weeks now on some of my DSC Resources using the standard unit and integration tests (from xNetworking) and it has been working well. See https://github.com/PlagueHO/cFSRM/tree/master/Tests for examples of these functions being used.

Thanks!

PlagueHO commented:
Awesome! Thanks for the feedback @KarolKaczmarek . I have updated the code.

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,209 @@ function Install-ResourceDesigner {
+    # tests are called. This is to ensure the correct module is tested.
+    [String] $OldModulePath = $env:PSModulePath
+    [String] $NewModulePath = $OldModulePath
+    if (($NewModulePath).Split(';') -ccontains $modulesFolder)
```
I still don't think this is correct. We add modulesFolder to the beginning if PSModulePath already contains it?

Something like this should suffice:
if (($env:PSModulePath.Split(';') | Select-Object -First 1) -ne $modulesFolder) 
{ 
     $env:PSModulePath = "$modulesFolder;$env:PSModulePath" 
}

tysonjhayes replied:
I think the way he wrote it is better, specifically it also covers the event that the folder name is contained in the ModulePath but is not the first one in particular. As it's possible the machine has the folder the user is working out of but is the first one.  This way we don't append more folders on then needed.

KarolKaczmarek replied:
My concern is that we only add the folder to the PSModulePath if it already exists there
if (($NewModulePath).Split(';') -ccontains $modulesFolder) 

Isn't the main goal here to add it if it's not in the PSModulePath? 
If it's already there, then moving it to the very first position is a good idea, but we don't have the most important scenario covered.

Unless it's covered somewhere else and I missed it?

PlagueHO replied:
I completely understand - I didn't quite engage with the requirement. The end result of this code fragment should be that the module folder should exist and always be the first one in the $Env:PSModulePath. I'll correct it now. Thanks again to you all for reviewing this.

Would it be helpful to also make the change you suggested in Issue #12 at the same time or leave that for a separate PR?

KarolKaczmarek replied:
Thanks!
Please leave #12 for another PR as that one will affect all existing tests and will require different level of testing

PlagueHO replied:
Absolutely! :+1:

KarolKaczmarek replied:
cool, let me run some manual tests and I'll merge it once it's ready.
In the meantime, can you take a look at #14

PlagueHO replied:
There is an issue with this code now. I'm looking into it, but it's causing my Integration tests to fail with this:
![2015-12-16_10-56-17](https://cloud.githubusercontent.com/assets/7589164/11825408/badb0e10-a3e3-11e5-86de-25ef6d6d39db.png)

Let me have a look and identify the problem. Unit tests are still fine.

tysonjhayes replied:
Look to see if you have multiple paths containing the resource in your $env:PSModulePath - it's why I originally wrote the code to remove out the current path we were running in and only work out of Program Files.

KarolKaczmarek replied:
As @tysonjhayes mentioned, you probably have this module in some other path added to PSModulePath.

Ideally we should not be overwriting users's module in Program Files with the development version. But it will cause issues like this since it will be discovered in multiple places. Not sure what would be the best way to solve it other than making sure that we don't have same module deployed in multiple places (which is far from ideal)

PlagueHO replied:
I figured it out - the problem is actually caused by the Issue #12. This is the resulting 
```powershell
$ENV:PSModulePath after running the code:
    [String] $OldModulePath = $env:PSModulePath
    [String] $NewModulePath = $OldModulePath
    if (($NewModulePath).Split(';') -ccontains $modulesFolder)
    {
        # Remove the existing module from the module path if it exists
        $NewModulePath = ($NewModulePath -split ';' | Where-Object {$_ -ne $modulesFolder}) -join ';'
    }
    $NewModulePath = "$modulesFolder;$NewModulePath"

    Write-Verbose -Verbose (`
        'Updating $ENV:PSModulesFolder to "{0}".' `
            -f $NewModulePath)   

    $env:PSModulePath = $NewModulePath
    [System.Environment]::SetEnvironmentVariable('PSModulePath',$NewModulePath,[System.EnvironmentVariableTarget]::Machine)
```

![2015-12-16_11-31-06](https://cloud.githubusercontent.com/assets/7589164/11826240/010d5c08-a3e9-11e5-895d-53eb9d932fe3.png)

I completely see what @tysonjhayes is saying now and understand why the original code you wrote moved the module to the Program Files folder - as it does eliminate the problem caused by Issue #12.

I could put a fix in to ensure the current path is also removed from the $ENV:PSModulesPath - but if Issue #12 is fixed then this isn't required.

KarolKaczmarek replied:
oh ok, it's clear now... 
well I guess you can go ahead and include fix for #12 here as well then. We will need to verify it doesn't break any of the existing tests in other repo though.

KarolKaczmarek replied:
actually, since you are not modifying meta.tests.ps1 here, can you send it as a separate PR?

PlagueHO replied:
Cool. Will do.

PlagueHO replied:
I've made the PR for issue #12. That PR should be merged before this one (PR #11).

KarolKaczmarek replied:
#12 is merged. @PlagueHO , can you confirm it fixed the above issue?

KarolKaczmarek replied:
I confirmed that the meta.tests.ps1 cleans the PSModulePath after it's run.

PlagueHO replied:
Awesome! Just checking now. Thanks heaps!

PlagueHO commented:
Done! Thanks :+1:

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,210 @@ function Install-ResourceDesigner {
+        $TestEnvironment = Inialize-TestEnvironment `
+            -DSCModuleName 'xNetworking' `
+            -DSCResourceName 'MSFT_xFirewall'
+            
```
this example lacks -TestType Unit
We could make it default value, but I think it's better to require user to specify it explicitly so that it's clear that this function works differently for different test types.

PlagueHO replied:
Fixed! Thanks!

KarolKaczmarek commented on TestHelper.psm1:
```diff
@@ -201,3 +204,210 @@ function Install-ResourceDesigner {
+    } 
+    
+    # The folder where this module is found
+    [String] $moduleRoot = $pwd
```
Was testing it on my machine and looks like due to this line we have an assumption that we will run those functions from the root folder (e.g. C:\resources\xTimeZone). 
I think it's a bit limiting since users may want to run just subset of tests while invoking them from e.g. C:\resources\xTimeZone\tests\Unit etc.

I'm not a big fan of $pwd since it depends on location from which script is called, I usually use:
```
$rootPath = Split-Path -Parent $MyInvocation.MyCommand.Path
```

PlagueHO replied:
Cool - I'll use this code from now on - good to know! I have changed the code.

Thanks

KarolKaczmarek commented:
Thanks a lot @PlagueHO for this change and adding all the modifications we asked about! It's merged now. I also kicked off appveyor builds for all repos and everything seems fine.

PlagueHO commented:
Fantastic! Thanks very much. I'll for rebuilds on all my projects and make sure everything is working nicely.

PlagueHO commented:
I can confirm all my builds are passing (after applying the temporary APpVeyor pester fix). This looks very good to me. Thanks again!
